src/rgw/rgw_role.cc

```diff
@@ -0,0 +1,220 @@
+  return rgw_put_system_obj(store, store->get_zone_params().roles_pool, id,
+                bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWRole::store_name(bool exclusive)
+{
+  RGWNameToId nameToId;
+  nameToId.obj_id = id;
+
+  bufferlist bl;
+  ::encode(nameToId, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().roles_pool, name,
+              bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWRole::create(bool exclusive)
+{
+  int ret;
+
+  /* check to see the name is not used */
+  ret = read_id(name, id);
+  if (exclusive && ret == 0) {
+    ldout(cct, 0) << "ERROR: name " << name << " already in use for role id "
+                    << id << dendl;
+    return -EEXIST;
+  } else if ( ret < 0 && ret != -ENOENT) {
+    ldout(cct, 0) << "failed reading role id  " << id << ": "
+                  << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  /* create unique id */
+  uuid_d new_uuid;
+  char uuid_str[37];
+  new_uuid.generate_random();
+  new_uuid.print(uuid_str);
+  id = uuid_str;
+
+  // Creation time
+  creation_date = boost::posix_time::to_iso_extended_string(
+                          boost::posix_time::microsec_clock::universal_time());
+
+  string pool_name = store->get_zone_params().roles_pool.name;
+  ret = store_info(exclusive);
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR:  storing info for in pool: " << pool_name << ": "
+                  << id << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  return store_name(exclusive);
+}
+
+int RGWRole::delete_obj()
+{
+  string pool_name = store->get_zone_params().roles_pool.name;
+
+  int ret = read_name();
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: finding role name in pool: " << pool_name << ": "
+                  << name << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  // Delete id
+  ret = rgw_delete_system_obj(store, store->get_zone_params().roles_pool, id, NULL);
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: deleting role id from pool: " << pool_name << ": "
+                  << id << ": " << cpp_strerror(-ret) << dendl;
+  }
+
+  // Delete name
+  ret = rgw_delete_system_obj(store, store->get_zone_params().roles_pool, name, NULL);
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: deleting role name from pool: " << pool_name << ": "
+                  << name << ": " << cpp_strerror(-ret) << dendl;
+  }
+
+  return ret;
+}
+
+int RGWRole::list()
+{
+  string pool_name = store->get_zone_params().roles_pool.name;
+
+  int ret = read_name();
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: finding role name in pool: " << pool_name << ": "
+                  << name << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  ret = read_info();
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: finding role id in pool: " << pool_name << ": "
+                  << id << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  return 0;
+}
```

Comment:
this function is just reading the role into memory, right? can we name it `init()` or `read()` instead?